GERTi/GERTiClient.lua

```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 	transmitInformation(neighbors[1]["address"], neighbors[1]["port"], "RetrieveAddress", origination, tier, serialTable)
 	local _, _, _, port, _, payload = event.pull(5, "modem_message")
 	if payload ~= nil then
 		return transmitInformation(sendingModem, port, payload)
 	end
 end
 
+handler["RegisterComplete"] = function (eventName, receivingModem, sendingModem, port, distance, code, targetMA, iResponse)
+	if receivingModem == targetMA then
+		-- This is us!
+		iAddress = iResponse
+	end
+	-- The code that goes in the 'else' depends on what happens to RegisterNode
+end
+
 handler["RETURNSTART"] = function (eventName, receivingModem, sendingModem, port, distance, code, tier)
 	-- Store neighbor based on the returning tier
-	return storeNeighbors(eventName, receivingModem, sendingModem, port, distance, tier)
+	storeNeighbors(eventName, receivingModem, sendingModem, port, distance, tier)
+end
+
+handler["ResolveAddress"] = function (eventName, receivingModem, sendingModem, port, distance, code, gAddress)
+	addWaitingAddress(gAddress, function(response)
+		-- this indentation is weird.
+		-- Anyway, this is the callback to forward the resolution
+		transmitInformation(sendingModem, port, "AddressResolution", gAddress, response)
+	end)
+end
+
+handler["AddressResolution"] = function (eventName, receivingModem, sendingModem, port, distance, code, gAddress, mtAddress)
+	if waitingAddress[gAddress] then
+		waitingAddress[gAddress](mtAddress)
+		waitingAddress[gAddress] = nil
+	end
 end
 
 local function receivePacket(eventName, receivingModem, sendingModem, port, distance, code, ...)
 	print(code)
 	-- Attempt to call a handler function to further process the packet
 	if handler[code] ~= nil then
-		return handler[code](eventName, receivingModem, sendingModem, port, distance, code, ...)
+		handler[code](eventName, receivingModem, sendingModem, port, distance, code, ...)
+	end
+	-- Just to make it clear, the return value system was a bad idea.
+end
+
+local function emergencyCalcIA(modemAddress)
+	local p = 0
+	for i = 1, 3 do
+		p = p + (modemAddress:byte(i) * (16 ^ i))
+	end
+	p = p % 4096
+	if p == 0 then p = 1 end
+	return p
+end
+-- This is used to replace resolveAddress if the MNC is unavailable.
+-- Otherwise, this is replaced during startup procedure.
+local function addressResolver(gAddress)
+	-- A gAddress is simply a number in the MNC (unless it has a ':' in which case it goes to GERTe)
+	local addr = tonumber(gAddress)
+	if not addr then return end
+	for _, v in ipairs(neighbors) do
+		if addr == emergencyCalcIA(v.address) then
+			return addr
+		end
 	end
 end
 
 -- Begin startup ---------------------------------------------------------------------------------------------------------------------------
 -- transmit broadcast to check for neighboring GERTi enabled computers
 if tunnel then
 	tunnel.send("AddNeighbor")
-	receivePacket(event.pull(1, "modem_message"))
```

Comment:
Unsure of why this is removed.

Reply:
There is code below it that replaces it's function for both modems and tunnels.

Reply:
Will review after requested changes are made.